In [106]:
import numpy as np 
import pandas as pd

In [107]:
df = pd.concat(map(pd.read_csv, ['Brazil.csv', 'germany.csv','germany1.csv', 
                                 'Portugal1.csv', 'Spain1.csv', 'Switzerland.csv',
                                'Switzerland1.csv', 'France1.csv',
                                'Brazil1.csv', 'Netherlands.csv', 
                                'Belgium.csv',  'England.csv',
                                'Argentina.csv', 'Uruguay.csv', 
                                'Senegal.csv', 'Ecuador.csv', 'CostaRica.csv',
                                'Mexico.csv', 'France.csv', 'Denmark.csv',
                                'Australia.csv', 'Serbia.csv',
                                'Cameroon.csv', 'Ghana.csv', 'Iran.csv',
                                'Poland.csv', 'SaudI.csv', 'Tunisia.csv', 
                                 'Japan.csv', 'Croatia.csv', 'Morocco.csv',
                                 'Spain2.csv', 'Portugal2.csv', 
                                 'England2.csv', 'Belgium2.csv', 'Uruguay2.csv',
                                'Canada2.csv', 'Netherlands2.csv', 'Saudi2.csv', 
                                'Croatia2.csv', 'Cameroon2.csv', 'Denmark2.csv',
                                'Costa2.csv', 'Morocco2.csv', 'Poland2.csv',
                                'Tunisia2.csv', 'Japan2.csv', 'Argentina2.csv', 
                                'Slovakia.csv', 'Iraq.csv', 'Jamaica.csv',
                                'Kuwait.csv', 'ElSalvador.csv', 'China.csv',
                                'Turkey.csv', 'Ukraine.csv', 'Bolivia.csv',
                                'Togo.csv', 'Togo.csv', 'Angola.csv', 
                                  'Italy.csv', 'Paraguay.csv', 
                                 'Russia.csv',
                                'Algeria.csv', 'Chile.csv', 'Yugoslavia.csv',
                                'Scotland.csv', 'Romania.csv', 'Greece.csv',
                                'Honduras.csv', 'Austria.csv', 'Bulgaria.csv', 
                                 'IvoryCoast.csv', 'SouthAfrica.csv', 'RepIreland.csv',
                                'Slovenia.csv', 'Czechoslovakia.csv', 'Hungary.csv', 
                                'Egypt.csv', 'Norway.csv', 'Peru.csv',
                                'NewZealand.csv', 'Bosnia.csv', 'Panama.csv','UAE.csv',
                                'Iceland.csv', 'USA.csv', 'USA1.csv',  'Nigeria.csv',
                                 'Colombia.csv']))



In [100]:
# 158 with wc 2022 teams

In [109]:
df.shape

(258, 18)

In [110]:
df.duplicated().value_counts()

False    257
True       1
dtype: int64

In [111]:
df = df.drop_duplicates()

In [112]:
df.shape

(257, 18)

In [113]:
df.corr()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  \
goals_z          1.000000  0.638557   0.160156      0.401719  0.225173   
xg_z             0.638557  1.000000   0.436815      0.688626  0.328847   
crosses_z        0.160156  0.436815   1.000000      0.662629  0.269739   
boxtouches_z     0.401719  0.688626   0.662629      1.000000  0.384468   
passes_z         0.225173  0.328847   0.269739      0.384468  1.000000   
progpasses_z     0.084953  0.147565   0.306065      0.281826  0.743808   
takeons_z        0.289001  0.369965   0.172152      0.263831  0.409830   
progruns_z       0.234152  0.396205   0.299010      0.412796  0.591734   
tackles_z        0.255914  0.368220   0.252085      0.409371  0.746054   
interceptions_z  0.212606  0.299788   0.265718      0.318847  0.757162   
clearances_z    -0.219942 -0.379485  -0.266693     -0.414176 -0.128940   
blocks_z         0.078048  0.084044   0.058444      0.149842  0.411209   
aerials_z        0.061632  0.088740   0.235791      0.251500 -0.030641   
fouls_z         -0.053061  0.019064   0.103259      0.066899  0.101253   
fouled_z         0.080535  0.208696   0.194502      0.209595  0.314279   
nsxg_z           0.427311  0.681841   0.713082      0.870455  0.487211   
results         -0.626918 -0.335499  -0.053791     -0.249146 -0.173075   

                 progpasses_z  takeons_z  progruns_z  tackles_z  \
goals_z              0.084953   0.289001    0.234152   0.255914   
xg_z                 0.147565   0.369965    0.396205   0.368220   
crosses_z            0.306065   0.172152    0.299010   0.252085   
boxtouches_z         0.281826   0.263831    0.412796   0.409371   
passes_z             0.743808   0.409830    0.591734   0.746054   
progpasses_z         1.000000   0.176909    0.332324   0.509211   
takeons_z            0.176909   1.000000    0.550408   0.411019   
progruns_z           0.332324   0.550408    1.000000   0.542149   
tackles_z            0.509211   0.411019    0.542149   1.000000   
interceptions_z      0.566163   0.399501    0.446074   0.714438   
clearances_z        -0.018902  -0.119814   -0.121749  -0.226415   
blocks_z             0.362509   0.193396    0.288748   0.299816   
aerials_z            0.038046  -0.108774   -0.073658   0.011667   
fouls_z              0.237940   0.129192    0.037505   0.029959   
fouled_z             0.192751   0.330894    0.268885   0.316676   
nsxg_z               0.395213   0.402790    0.492808   0.482706   
results             -0.052657  -0.193502   -0.219373  -0.217288   

                 interceptions_z  clearances_z  blocks_z  aerials_z   fouls_z  \
goals_z                 0.212606     -0.219942  0.078048   0.061632 -0.053061   
xg_z                    0.299788     -0.379485  0.084044   0.088740  0.019064   
crosses_z               0.265718     -0.266693  0.058444   0.235791  0.103259   
boxtouches_z            0.318847     -0.414176  0.149842   0.251500  0.066899   
passes_z                0.757162     -0.128940  0.411209  -0.030641  0.101253   
progpasses_z            0.566163     -0.018902  0.362509   0.038046  0.237940   
takeons_z               0.399501     -0.119814  0.193396  -0.108774  0.129192   
progruns_z              0.446074     -0.121749  0.288748  -0.073658  0.037505   
tackles_z               0.714438     -0.226415  0.299816   0.011667  0.029959   
interceptions_z         1.000000     -0.147269  0.333417  -0.002968  0.147834   
clearances_z           -0.147269      1.000000  0.189945   0.280982  0.215660   
blocks_z                0.333417      0.189945  1.000000   0.032541  0.198551   
aerials_z              -0.002968      0.280982  0.032541   1.000000  0.187114   
fouls_z                 0.147834      0.215660  0.198551   0.187114  1.000000   
fouled_z                0.320693      0.009436  0.099659   0.018854  0.275511   
nsxg_z                  0.434678     -0.419465  0.183517   0.173429  0.115044   
results                -0.141502      0.111652 -0.101842  -0.091168  0.071193   

                 fo

# convert to csv

In [114]:
df.to_csv('./WC Data Models/WC-stats.csv', index=False)


In [121]:
from sklearn.linear_model import LinearRegression
from sklearn.inspection import permutation_importance
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge



In [70]:
df.head()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0  0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1  0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2  0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3  0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4  0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   

   takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0   0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1   0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2   0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3   0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4   0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   

   aerials_z   fouls_z  fouled_z    nsxg_z  results  
0  -0.020000 -0.224615  0.100769 -0.124615        2  
1   0.196923 -0.016923  0.033846  0.146923        1  
2   0.683846  0.213846  0.353846  0.786923        9  
3   0.038462  0.039231  0.022308  0.692308        5  
4   0.572308 -0.016154 -0.096923  0.890769        5

In [16]:
y = df['results']

In [17]:
X = df.drop(columns=['results'])

In [18]:
X_train, X_test, y_train, y_test = train_test_split(
     X, y, test_size=0.20, random_state=42)

In [19]:
model = Ridge(alpha=1e-2).fit(X_train, y_train)
model.score(X, y)

0.4012771300488499

In [20]:
r = permutation_importance(model, X, y, n_repeats=30, random_state=0)

# Permutation feature importance

In [21]:
scoring = ['r2', 'neg_mean_absolute_percentage_error', 'neg_mean_squared_error']
r_multi = permutation_importance(
     model, X, y, n_repeats=30, random_state=0, scoring=scoring)

for metric in r_multi:
     print(f"{metric}") 
     r = r_multi[metric]
     for i in r.importances_mean.argsort()[::-1]:
            if r.importances_mean[i] - 2 * r.importances_std[i] > 0:
                print(f"    {X.columns[i]:<8}"
                      f"{r.importances_mean[i]:.3f}"
                      f" +/- {r.importances_std[i]:.3f}")
                
                

r2
    goals_z 0.891 +/- 0.089
    xg_z    0.056 +/- 0.017
    fouled_z0.029 +/- 0.011
    progruns_z0.029 +/- 0.010
    crosses_z0.012 +/- 0.006
neg_mean_absolute_percentage_error
    goals_z 0.659 +/- 0.066
    fouled_z0.039 +/- 0.016
    progruns_z0.028 +/- 0.012
    fouls_z 0.023 +/- 0.011
    tackles_z0.017 +/- 0.006
neg_mean_squared_error
    goals_z 66.643 +/- 6.680
    xg_z    4.222 +/- 1.289
    fouled_z2.158 +/- 0.859
    progruns_z2.150 +/- 0.757
    crosses_z0.932 +/- 0.433


- df.shape = (158, 17)
- Ridge Regression = 0.42629810615686115
- r2
    - goals_z 1.282 +/- 0.149
    - progruns_z0.096 +/- 0.029
    - xg_z    0.095 +/- 0.030
    - takeons_z0.026 +/- 0.013
- neg_mean_absolute_percentage_error
    - goals_z 0.988 +/- 0.115
- neg_mean_squared_error
    - goals_z 108.981 +/- 12.672
    - progruns_z8.139 +/- 2.494
    - xg_z    8.047 +/- 2.509
    - takeons_z2.195 +/- 1.067


- df.shape = (125, 17)

- Ridge Regression = 0.4494483686544971

- r2
    - goals_z 0.982 +/- 0.123
    - progruns_z0.132 +/- 0.039
    - xg_z    0.056 +/- 0.017
    - boxtouches_z0.046 +/- 0.023
    - nsxg_z  0.039 +/- 0.013
    - aerials_z0.031 +/- 0.014
- neg_mean_absolute_percentage_error
    - goals_z 0.888 +/- 0.122
    - progruns_z0.181 +/- 0.059
- neg_mean_squared_error
    - goals_z 89.125 +/- 11.206
    - progruns_z11.964 +/- 3.505
    - xg_z    5.055 +/- 1.543
    - boxtouches_z4.180 +/- 2.062
    - nsxg_z  3.509 +/- 1.182
    - aerials_z2.782 +/- 1.244

- df.shape = (145, 17)
- Ridge Regression = 0.43985733944194594

- r2
    - goals_z 0.919 +/- 0.122
    - progruns_z0.077 +/- 0.028
    - nsxg_z  0.069 +/- 0.026
    - boxtouches_z0.034 +/- 0.015
- neg_mean_absolute_percentage_error
    - goals_z 0.817 +/- 0.115
    - progruns_z0.124 +/- 0.036
    - tackles_z0.060 +/- 0.022
    - boxtouches_z0.048 +/- 0.023
- neg_mean_squared_error
    - goals_z 78.467 +/- 10.375
    - progruns_z6.549 +/- 2.364
    - nsxg_z  5.854 +/- 2.186
    - boxtouches_z2.943 +/- 1.292

- NMAPE = the higher the better
- NMSE
- R2 =  is a statistical measure of how close the data are to the fitted regression line . It is the percentage of variation (from 0 to 1) explained by the relationship between two variables.. metric that is used to evaluate the performance of a regression-based machine learning model.